### PR TITLE
refactor(#255): cmd/mod.rs complexity reduction

### LINKED ISSUE #255: tech-debt: cmd/mod.rs complexity reduction (scrub_db_args, run_parsed_command_with_mode, file splitting)
## Context

Identified during PR #254 code review (Issue #214: Parser Infrastructure Cleanup). These are pre-existing issues in code touched but not introduced by the PR.

### 1. scrub_db_args complexity
- **File**: `crates/rskim/src/cmd/mod.rs:1024`
- **Issue**: 77 lines, 4 nesting levels, manual index loop
- **Suggested fix**: Extract `classify_token` helper with `TokenAction` enum
- **Risk**: Touches security-sensitive arg scrubbing logic — needs careful testing

### 2. run_parsed_command_with_mode mixed concerns
- **File**: `crates/rskim/src/cmd/mod.rs:502`
- **Issue**: 86 lines with 7 mixed concerns (spawn, capture, parse, format, record, stats, passthrough)
- **Suggested fix**: Extract passthrough guard and analytics recording into separate helpers

### 3. cmd/mod.rs file length
- **File**: `crates/rskim/src/cmd/mod.rs`
- **Issue**: 1828 lines (3.6x the 500-line threshold)
- **Suggested fix**: Split into submodules: `dispatch.rs`, `runner.rs`, `security.rs`, `test_support.rs`
- **Risk**: Requires careful re-export management across ~60 consumer files

## Source
- Review: `.devflow/docs/reviews/feature-214-parser-infrastructure-cleanup/2026-05-26_0958/`
- PR: #254

### PR DESCRIPTION
**Pull request by dean0x**
## Summary

- Split `crates/rskim/src/cmd/mod.rs` (2,070 lines) into 5 focused submodules: `security.rs`, `registry.rs`, `execution.rs`, `dispatch.rs`, `test_support.rs`
- Refactored `scrub_db_args` with `TokenAction` enum + `classify_token` helper — eliminates 4-level nesting with a flat match
- Decomposed `run_parsed_command_with_mode` into orchestrator + `passthrough_raw` + `record_and_report` helpers
- All `pub(crate)` items re-exported from `mod.rs` — zero consumer files modified

## Details

**mod.rs reduction:** 2,070 → 420 lines (80% reduction). Retains module declarations, re-exports, shared utilities (stdin/passthrough/flag parsers).

**New modules:**
| Module | Lines | Responsibility |
|--------|-------|----------------|
| `security.rs` | 494 | Credential scrubbing (`scrub_db_args`, `sanitize_for_display`, `TokenAction` enum) |
| `registry.rs` | 288 | Subcommand registry (`KNOWN_SUBCOMMANDS`, `META_SUBCOMMANDS`, `wrapper_targets`) |
| `execution.rs` | 466 | Execution pipeline (`RunContext`, `run_parsed_command_with_mode`, `run_tool`, `combine_output`) |
| `dispatch.rs` | 576 | Top-level routing (`dispatch`, `dispatch_cargo/go/swift/dotnet`, `extract_subcmd`) |
| `test_support.rs` | 69 | Test helpers (`make_output`, `load_fixture`) — `#[cfg(test)]` gated |

**API preservation:** All existing import paths (`crate::cmd::dispatch()`, `super::RunContext`, `crate::cmd::sanitize_for_display`) continue to work via re-exports. Zero files outside `crates/rskim/src/cmd/` were modified.

**New tests:** 9 `classify_token` unit tests covering all `TokenAction` variants including edge cases (`-P` as port not password, URI without `@`, empty string).

Closes #255

## Test plan

- [x] `cargo test -p rskim --all-features` — 4,893 tests pass (baseline 3,631 + 9 new classify_token)
- [x] `cargo clippy -p rskim -- -D warnings` — zero warnings
- [x] `cargo fmt -p rskim -- --check` — clean
- [x] `cargo test -p rskim --test 'cli_e2e_*'` — 387 integration tests pass
- [x] `git diff --name-only main...HEAD` — only `crates/rskim/src/cmd/` files
- [x] No `unsafe`, no new dependencies, no `#[allow(...)]` except one justified `too_many_arguments` on `record_and_report`

### CODE REVIEW

**Comment by dean0x** [comment]
## Code Review — Cycle 1

**Recommendation: APPROVED_WITH_CONDITIONS** | Average Score: 8.6/10

9 reviewers (security, architecture, performance, complexity, consistency, regression, testing, reliability, rust). All 4,893 tests pass. Zero consumer files modified. Zero behavioral changes.

### Blocking (1)

| ID | Finding | File | Confidence |
|----|---------|------|------------|
| F-1 | `record_and_report` has 8 positional params with `#[allow(clippy::too_many_arguments)]` — bundle into struct | `execution.rs:202` | 85% (3 reviewers) |

### Should-Fix (2)

| ID | Finding | File | Confidence |
|----|---------|------|------------|
| F-2 | New `mod` declarations scattered instead of grouped at top | `mod.rs:181+` | 82% |
| F-3 | `debug_assert!` in `prepend_without` compiles away in release — use `assert!` | `dispatch.rs:30` | 85% |

### Suggestions (60-79% confidence, non-blocking)

- `TokenAction` enum uses owned `String` where `&str` borrows suffice (`security.rs:36`, 75%)
- `format_analytics_label` only scrubs credentials for `"db"` family — pre-existing, follow-up ticket candidate (`execution.rs:331`, 82%)
- Dangling intra-doc link to `[WRAPPER_TARGETS]` in `mod.rs:97` (85%)
- Missing `combine_output` unit test (`execution.rs`, 70%)

### What Went Well

- Clean SRP decomposition: 2,070 → 420 lines (80% reduction)
- `TokenAction` enum + `classify_token` eliminates 4-level nesting
- `pub(crate) use` re-export strategy preserves all import paths
- 9 new `classify_token` tests with edge case coverage
- Sync-guard tests protect sorted invariants and dispatch coverage
- Zero regression risk (regression reviewer: 10/10)

Full review: `.devflow/docs/reviews/refactor-255-cmd-mod-complexity/2026-06-02_1034/`

File: crates/rskim/src/cmd/execution.rs [hunk 1]
```diff
@@ -0,0 +1,466 @@
+//! Command execution infrastructure for skim CLI.
```
Comment (dean0x) [inline crates/rskim/src/cmd/execution.rs:1]: **[F-1]** `record_and_report` takes 8 positional params with clippy suppression. Bundle into a struct to match the PR pattern used for ParsedCommandConfig/ToolRunConfig. (85%, 3 reviewers)

**Review by dean0x — COMMENTED** [review]
See review summary comment above for full details.

**Comment by dean0x** [comment]
## Security Issue: Credential-bearing URLs not redacted

**File**: crates/rskim/src/cmd/security.rs
**Function**: scrub_infra_args (line 190)
**Confidence**: 85%

The `scrub_infra_args` function redacts flag-based credentials (--token, --password, -H Authorization headers) but does NOT scrub credential-bearing URLs like `https://TOKEN@host/repo` or `https://user:pass@registry`.

The non-streaming path through `run_tool` → `format_analytics_label("infra", ...)` relies solely on `scrub_infra_args`, which lacks URL credential handling. This would persist tokens in URLs to `analytics.db`.

The streaming infra path correctly applies `scrub_credential_url` pre-scrubbing. The non-streaming path needs the same coverage.

### Suggested Fix
Apply `scrub_credential_url` to each token in `scrub_infra_args`, or have `format_analytics_label` always scrub credential URLs for ALL families (not just "db" and "infra").

**Comment by dean0x** [comment]
## Code Complexity: Header-handling nesting depth

**File**: crates/rskim/src/cmd/security.rs
**Block**: lines 229-247 (header value handling)
**Confidence**: 85%

The `-H`/`--header` handling block nests 5 levels deep:
- `while > if(tok=="-H") > if(i < tokens.len()) > if(lower.starts_with("authorization:")) > while(i < tokens.len())`

This exceeds the warning threshold (>4 levels). The inner continuation-scanning loop is particularly hard to follow because it silently advances `i` without pushing to `out`.

### Suggested Fix
Extract into a helper function `scrub_header_value(tokens: &[&str], i: usize, out: &mut Vec<String>) -> usize` that encapsulates the header classification and continuation-token scanning. This would flatten nesting by one level and isolate the logic for testing.

**Comment by dean0x** [comment]
## Architecture: String-based dispatch pattern

**File**: crates/rskim/src/cmd/execution.rs  
**Function**: format_analytics_label (line 344)
**Confidence**: 82%

The function dispatches scrubbing via `if family == "db" / else if family == "infra"` string comparisons. Adding a third family needing scrubbing requires modifying this function (violates Open/Closed Principle). With security-critical scrubbing, the risk of missing a branch is real.

### Suggested Fix
Use a match expression instead:
```rust
let scrubbed_rest = match family {
    "db" => scrub_db_args(rest),
    "infra" => scrub_infra_args(rest),
    _ => rest.to_string(),
};
format!("skim {family} {program} {scrubbed_rest}")
```
This also eliminates the separate `rest.is_empty()` early return.

**Comment by dean0x** [comment]
## Code Consistency: Asymmetric scrubbing patterns

**File**: crates/rskim/src/cmd/security.rs
**Functions**: scrub_db_args vs scrub_infra_args
**Confidence**: 85%

`scrub_db_args` was refactored to use a clean `TokenAction` enum + `classify_token` pure-function pattern (excellent design), but `scrub_infra_args` in the same file uses an inline `while` loop with ad-hoc `if/continue` chains.

Two scrubbing functions in the same module follow different architectural patterns. This creates inconsistency and makes the codebase harder to maintain.

### Suggested Fix
Apply the same `TokenAction`-style decomposition to `scrub_infra_args`. Introduce an `InfraTokenAction` enum or extend `TokenAction` with infra variants like `RedactHeader`. This aligns both scrubbers with the established pattern and makes them independently testable at the classification level.

**Comment by dean0x** [comment]
## Code Review Summary: PR #267 (refactor/255-cmd-mod-complexity)

### Overview
This PR achieves excellent modular decomposition: the 2,070-line `cmd/mod.rs` god module is split into 5 focused modules with clear single responsibilities. The refactoring is well-executed with **zero regressions** and **no consumer changes needed** (all re-exported via `mod.rs`).

**Architecture Score**: 8/10  
**Security Score**: 8/10  
**Complexity Score**: 8/10

---

### High-Confidence Findings (≥80%)

#### Blocking Issues
1. **Security: Credential URLs not scrubbed in infra path** (85%) — [Detailed comment above] — The non-streaming `run_tool` path through `format_analytics_label("infra", ...)` relies solely on `scrub_infra_args`, which lacks URL credential handling. Tokens in URLs like `https://TOKEN@host/repo` would persist to `analytics.db`.

#### Should-Fix Issues
2. **Complexity: Header-handling nesting (5 levels)** (85%) — [Detailed comment above] — Extract header classification into a helper function to flatten nesting and improve testability.

3. **Architecture: String-based family dispatch** (82%) — [Detailed comment above] — Use match expression instead of if-chain for OCP compliance and scalability.

4. **Consistency: scrub_infra_args lacks TokenAction pattern** (85%) — [Detailed comment above] — Apply the same elegant `TokenAction` + `classify_token` decomposition that `scrub_db_args` uses.

---

### Lower-Confidence Suggestions (60-79%)

These should be considered but are not blocking:

- **`-h` flag over-redaction in scrub_db_args** (82%) — The `-h` flag is DB-tool-specific (means `--host` for psql/mysql). Add a doc comment noting this function is DB-family-specific and should not be reused for other families without reviewing flag semantics.

- **dispatch.rs help printers could be data-driven** (65%) — Four `print_*_help()` functions contain nearly identical structure (header, usage, subcommands, examples). A data-driven help builder would reduce boilerplate, but this is a style preference given the current count of 4 dispatchers.

- **scrub_infra_args header-continuation heuristic** (70%) — The loop at lines 241-247 stops at `-`-prefixed tokens or `http://`/`https://` URLs. A header value containing a URL-like string that is not a flag boundary could cause under-redaction. Worth a clarifying comment.

- **CommandRunner inline instantiation** (80%, pre-existing) — `run_raw_passthrough` and `obtain_output` both construct `CommandRunner::new(...)` inline rather than accepting it as a parameter. Pre-existing pattern; low priority for future refactoring.

---

### Positive Observations

✅ **Clean decomposition**: Each module has one reason to change (dispatch routing, execution, registry, security, test helpers)

✅ **Acyclic dependencies**: No circular imports; clean DAG between submodules

✅ **Deep modules (Ousterhout principle)**: Small public API surface, substantial implementation complexity hidden

✅ **TokenAction enum**: Converts deeply nested if-chains into pure function returning discriminated union — textbook complexity reduction

✅ **Parameter bundling**: `ParsedCommandConfig`, `ToolRunConfig`, `RunContext`, `RecordReport` replace positional parameter lists with well-documented named structs

✅ **Zero consumer changes**: All 40+ call sites at `crate::cmd::*` continue working without modification

---

### Recommendation

**APPROVED WITH CONDITIONS**

Merge after addressing:
1. **[BLOCKING]** Add URL credential scrubbing to `scrub_infra_args` (security gap)
2. **[SHOULD-FIX]** Refactor `format_analytics_label` to use match instead of if-chain
3. **[SHOULD-FIX]** Extract header handling into helper function to reduce nesting depth

The lower-confidence suggestions (`-h` over-redaction, data-driven help, header heuristic clarity) can be addressed in a follow-up if desired.

---

**Reviewed by**: Claude Code  
**Branch**: refactor/255-cmd-mod-complexity → main  
**Files**: 6 (dispatch.rs, execution.rs, mod.rs, registry.rs, security.rs, test_support.rs)  
**Changes**: +2180 -1672
